File: app-ios-shared/src/commonMain/kotlin/io/github/droidkaigi/confsched/shared/share/ShareNavigator.kt
package io.github.droidkaigi.confsched.shared.share

import androidx.compose.ui.graphics.ImageBitmap
import co.touchlab.kermit.Logger
import io.github.droidkaigi.confsched.droidkaigiui.toUiImage
import platform.UIKit.UIActivityViewController
import platform.UIKit.UIApplication

class ShareNavigator {
    fun share(text: String) {
        val items = mutableListOf<Any>(text)

        val activityViewController = UIActivityViewController(items, null)
        val keyWindow = UIApplication.sharedApplication.keyWindow
        val rootViewController = keyWindow?.rootViewController

        rootViewController?.presentViewController(
            viewControllerToPresent = activityViewController,
            animated = true,
            completion = null,
        )
    }

    fun shareTextWithImage(text: String, image: ImageBitmap) {
        val items = mutableListOf<Any>(text)
        image.toUiImage()?.let {
            items.add(it)
            Logger.d("Success to create UIImage from ImageBitmap.")
        } ?: run {
            Logger.d("Failed to create UIImage from ImageBitmap.")
        }

        val activityViewController = UIActivityViewController(items, null)
        val keyWindow = UIApplication.sharedApplication.keyWindow
        val rootViewController = keyWindow?.rootViewController

        rootViewController?.presentViewController(
            viewControllerToPresent = activityViewController,
            animated = true,
            completion = null,
        )
    }
}

File: app-ios/Sources/CommonComponents/ShareNavigator.swift
import UIKit

public struct ShareNavigator {
    public init() {}
    
    @MainActor
    public func shareTextWithImage(text: String, image: UIImage) {
        let items: [Any] = [text, image]
        let activityViewController = UIActivityViewController(activityItems: items, applicationActivities: nil)
        
        DispatchQueue.main.async {
            if let windowScene = UIApplication.shared.connectedScenes.first as? UIWindowScene,
               let keyWindow = windowScene.windows.first(where: { $0.isKeyWindow }),
               let rootViewController = keyWindow.rootViewController {
                rootViewController.present(activityViewController, animated: true, completion: nil)
            } else {
                print("Unable to find the root view controller.")
            }
        }
    }
}
